### ISSUE TITLE
Rpc session on user-provided connection

### ISSUE THREAD

**Issue by elbaro** [google/tarpc#217]
Can one create a rpc client/server on a user provided connection (e.g. std::net or tokio TcpStream/UnixStream)? I would like to have more control on connections and do these things:

- multiplexing, serve different RPC servers
```rs
let stream = TcpStream::connect("127.0.0.1:555".parse().unwrap()).unwrap();
let plex = some_multiplexer::new(stream);
let conn1 = plex.open();
let conn2 = plex.open();
let rpc_client1 = RpcClient1::new(conn1);
let rpc_client2 = RpcClient2::new(conn2);
```

- Gracious exit with stream/future
```rs
use stream_cancel::StreamExt;

let listener = tokio::net::UnixListener::bind("/tmp/example.sock").unwrap();
listener
	.incoming()
	.take_until(is_terminated)
	.for_each(move |stream| {
		let stream = stream.take_until(is_terminated);
		let rpc_server = RpcServer::new(stream)
                rpc_server.serve_calls();
	});

exit(); // is_terminate resolves to completed
```

- Built-in policy like ConnectionFilter can be customized

**Comment by tikue** [comment]
Hi! I think you should be able to do some of these things today.

For multiplexing, you could either provide a multiplexing transport, possibly by multiplexing across multiple [channel transports](https://github.com/google/tarpc/blob/master/rpc/src/transport/channel.rs); or you could implement a service that multiplexes across other services, like I do in [service registry](https://github.com/google/tarpc/blob/4e0be5b626dc940e2c4b9dd61e87ab0302113938/tarpc/examples/service_registry.rs#L377).

For gracious exit, you can use the server's connection stream to decide when to stop. I do something like that in the [readme example](https://github.com/google/tarpc/blob/4e0be5b626dc940e2c4b9dd61e87ab0302113938/tarpc/examples/readme.rs#L60). For closing individual connections, I believe you can select the future representing the connection with your exit signal. So instead of `incoming.respond_with(serve(server))` you can do `incoming.for_each(|channel| select(
    channel.respond_with(serve(server.clone())), 
    exit_signal
))`.

Can you say more about how you'd like to customized ConnectionFilter? That's a built-in policy, like you say, but you can provide your own filters on the connection stream, similar to how you'd hook in gracious exit.

**Comment by elbaro** [comment]
Thanks for the answer. Basically I want lower-level channel API rather than `Server` struct.

Since a server-side push is out-of-scope (#39), I'd like to have two transports with the opposite roles of client/server. Fortunately I found [yamux](https://github.com/paritytech/yamux) which allows me to write multiplexing TcpConnections in a few lines. This also enables us to have other protocols along with tarpc.

A service registry does not help in this case. That is one-directional. Injecting my stream combinator has two problems:

```
let tcp_conn_stream = ..;
let transport = multiplex(tcp_conn_stream); // stream of (conn1, conn2, conn3)

let server = Server::default()
        .incoming(transport)  // 1. only want to use (conn1, conn2) and reserve conn3 for HTTP
                              // 2. incoming accepts a stream of tarpc::server::Channel
                              // but we want conn1 for server::Channel and conn2 for Client

        .for_each(|channel| { // type server::channel already implies one-directional
            channel.respond_with(serve(HelloServer))
        })
```

You can see `Server` API is good as a high-level API but does not allow fine control. Currently server::Channel is a little tied to Server struct (client_addr, closed_connections, etc). Maybe exposing this as a low-level API (server::Channel::from(tcp).serve(..)) would be enough.

For gracious exit and ConnectionFilter, you are right. I was just listing things that can be easily done with low-level API.

**Comment by elbaro** [comment]
I made some modification for my use case ([here](https://github.com/elbaro/tarpc/tree/transport-patch)) and it is working well.
To work with general `AsyncRead+AsyncWrite` other than `TcpStream`, [a dummy implementation](https://github.com/elbaro/tarpc/blob/135e289c6efc7465bdd9cf490afd32adab3c8164/bincode-transport/src/lib.rs#L109-L125) of `rpc::Transport` is added.
Also to create a channel from a given Transport without `tarpc::Server`, a channel [constructor](https://github.com/elbaro/tarpc/blob/93402c5400782a616033634029da453a81ce2c73/rpc/src/server/mod.rs#L223-L239) is required.

These are temporary solution because `peer_addr` or `local_addr` doesn't make sense for non-TCP protocols like UDS or yamux.

**Comment by tikue** [comment]
I'm basically unopposed to factoring out the `peer_addr()` and `local_addr()` methods. Tracing can use other things, and the filter stuff can probably be an opt-in policy, like maybe just chained on the connection stream. Would you be interested in putting together a PR for that?

I'd have to take a closer look at the other changes to fully understand.

**Comment by elbaro** [comment]
Can you explain how to acheive bi-directinoal RPC with `Incoming`+`tarpc::Server`?

This is how I implemented with channels:
https://github.com/elbaro/grass/blob/197d34a33cc957e4086dfa928051b8ac6e162d0f/src/broker.rs#L96-L143

**Comment by tikue** [comment]
So I got some time to think about what you said about `ConnectionFilter`. I have a [branch](https://github.com/google/tarpc/tree/filter-refactor) that tears out the `peer_addr` and `local_addr` methods on `Transport` and makes `ConnectionFilter` optional by providing a key to filter connections on:

```rust
    // The server is configured with the defaults.
    let server = Server::default()
        // Server can listen on any type that implements the Transport trait.
        .incoming(transport)
        // Connection limits use the default configuration, keyed on IP.
        .limit_connections(
            {
                let mut config = server::filter::Config::default();
                config.max_connections_per_key = 1;
                config.max_connections = 10;
                config
            },
            |t| t.peer_addr().unwrap().ip(),
        )
        // serve is generated by the service! macro. It takes as input any type implementing
        // the generated Service trait.
        .respond_with(service::serve(HelloServer));
```

I'd love it if you could test out the branch and see if it works for you! I feel positively about this change, though it is a backwards compatibility hazard: anyone who updates to this version of tarpc will have to opt-in to some defenses that previously were default-enabled.

**Comment by elbaro** [comment]
Thanks for the commits. I reviewed the code and agree that this is better. However my use case (as in the link above) is not addressed by this change,  but blocked by the limitation of using tarpc::Server.

A question-
-Is tarpc going to support general `AsyncRead+AsyncWrite`, or is tarpc opinionated about how connections are made for the benefits of providing easy APIs?

If the former, It still would be great if tarpc exposes the Channel constructor. I think it is impossible anyway for tarpc::Server to cover all possible scenarios of creating/handling connections. (see my link above) tarpc::Server should be a helper class that fits 90% of use cases but not 100%.

Another blocker for using `AsyncRead+AsyncWrite` with tarpc is `rpc::Transport` because it is only implemented for TCP.  Can it be replaced by Sink+Send? [bincode-transport](https://github.com/google/tarpc/blob/master/bincode-transport/src/lib.rs) looks like just a wrapper for TcpStream.

**Comment by tikue** [comment]
Thanks for your comments. I'd like to dig in a bit to why you prefer `AsyncRead+AsyncWrite` over `Transport`? `Transport` can be implemented by anyone for any protocol, not just TCP. In fact it's now just an alias for `Stream+Sink`.

**Comment by elbaro** [comment]
AsyncRead+AsyncWrite is more generic than Transport.
Users can add Transport, and it is not hard, but how about unifying them with something like

```
impl<Item, SinkItem> rpc::Transport for bincode_transport::Transport<A, Item, SinkItem>
    where A: AsyncRead+AsyncWrite
```

instead of 

```
impl<Item, SinkItem> rpc::Transport for bincode_transport::Transport<TcpStream, Item, SinkItem>
impl<Item, SinkItem> rpc::Transport for bincode_transport::Transport<UnixStream, Item, SinkItem>
...
impl<Item, SinkItem> rpc::Transport for json_transport::Transport<TcpStream, Item, SinkItem>
impl<Item, SinkItem> rpc::Transport for json_transport::Transport<UnixStream, Item, SinkItem> {
    fn peer_addr() -> &str { "0.0.0.0" } // unused code
}
...
```

This is not really a blocker but it doesn't feel right that users have to add such dummy code to use tarpc.

**Comment by tikue** [comment]
`AsyncWrite+AsyncRead` can't release in memory transports that don't serialize requests.

**Comment by tikue** [comment]
You may want to take a look at my open PR. It removes the transport trait so users don't have to impl it.

**Comment by tikue** [comment]
@elbaro did you still have any open questions?

**Comment by donggu-prestolabs** [comment]
I will take a look at new releases and reopen if required.

**Comment by elbaro** [comment]
I will take a look at new releases and reopen if required.
